CI: E2E jobs collide on spiced's fixed ports 8090/50051 on the shared macOS pool, dequeuing PRs with only a timeout to show for it

## Summary

Every E2E job that runs `spice run` binds spiced's **default** ports — `127.0.0.1:8090` (HTTP) and
`127.0.0.1:50051` (Flight). `e2e_test_ci.yml` schedules many such jobs onto the `spiceai-macos`
label, and that pool runs **several runners per physical host** (`/opt/github-runner-01`,
`-02`, `-03`, … on the same machine), which share one network namespace. When two of those jobs
overlap on one host, the second `spice run` cannot bind, exits, and the job's

```
while [[ "$(curl -s http://localhost:8090/v1/ready)" != "ready" ]]; do sleep 1; done
```

loop polls for the full `timeout-minutes` before failing with nothing but a timeout message. In
the merge queue that dequeues the PR.

No diff can cause this and no author can fix it.

## Signature

Run summary:

```
##[error]The action 'Wait for Spice runtime is ready' has timed out after 5 minutes.
```

`spice.log`, dumped later by `Stop spice and check logs`:

```
INFO  runtime::init::model: Model [hf_local_model] deployed, ready for inferencing
INFO  runtime: Shutdown initiated; waiting up to 30s for connections to drain
ERROR spiced: Unable to start Spice Runtime servers: Unable to start HTTP server:
      Unable to bind to address: Address already in use (os error 48)
INFO  spiced: Goodbye!
```

and, on the sibling job of the same run, the Flight port instead:

```
ERROR spiced: Unable to start Spice Runtime servers: Unable to start Flight server:
      Address 127.0.0.1:50051 is already in use by another process.
```

Note the ordering: the runtime loads *everything* successfully — datasets registered, model
deployed — and only then discovers it cannot serve. So the log looks healthy right up to the last
two lines.

## Blast radius

Window: all 74 `e2e_test_ci.yml` failures from 2026-07-27 to 2026-08-03T21:23Z.

Confirmed occurrences of this cause, all `event=merge_group`:

| Run | Created | Job(s) | Port that was taken |
|---|---|---|---|
| [30717979872](https://github.com/spiceai/spiceai/actions/runs/30717979872) | 2026-08-01T23:17Z | `openai model (darwin-aarch64)` | `50051` (Flight) |
| [30717979872](https://github.com/spiceai/spiceai/actions/runs/30717979872) | 2026-08-01T23:17Z | `huggingface model (darwin-aarch64)` | `8090` (HTTP) |
| [30798700215](https://github.com/spiceai/spiceai/actions/runs/30798700215) | 2026-08-03T10:26Z | `huggingface model (darwin-aarch64)` | `8090` (HTTP) |

Run 30717979872 is the clearest evidence: **two model jobs of the same run failed nine seconds
apart, each reporting the other's port as taken.** They were racing each other on one host.

Two further failures of the same step on 2026-08-03 (20:46Z, 21:03Z) are a *different* root cause
— a Hugging Face `401` on the embedding download — filed separately. Same symptom, unrelated
mechanism; worth stating explicitly so neither fix is mistaken for covering both.

## The quieter risk: this can also pass wrongly

If the process already holding `8090` is a *ready* runtime — a leftover from an earlier job on the
same host, or a concurrent job that got there first — then `curl http://localhost:8090/v1/ready`
returns `ready` immediately and the wait step **succeeds**. The tests then run against a runtime
loaded from a different spicepod. That is a silent wrong result rather than a red build, and it is
the same root cause. `Stop spice and check logs` does `pkill -x spice`, which also means a job
finishing on a shared host can kill a *sibling* job's runtime mid-test.

## Suggested direction

1. **Give each job its own ports.** `spiced` already supports `--http <BIND_ADDRESS>` and
   `--flight <BIND_ADDRESS>` (`crates/runtime/src/config.rs`), and `spice run` forwards trailing
   args (`trailing_var_arg = true`), so `spice run -- --http 127.0.0.1:$PORT --flight
   127.0.0.1:$FLIGHT_PORT` is available today. Derive the ports from something unique per job
   (e.g. `GITHUB_RUN_ID`/`GITHUB_JOB` hashed into a high range) and use the same variable in the
   readiness poll and every subsequent `curl`/test invocation. There are 26 hardcoded
   `localhost:8090` / `127.0.0.1:8090` references in `e2e_test_ci.yml` and 19 copies of the
   readiness loop, so this wants a small composite action rather than 19 edits.
2. **Fail fast and name the cause.** The readiness wait should notice that the `spice run` child
   has exited (it is started as `spice run &> spice.log &`, so its PID is available) and fail
   immediately with the tail of `spice.log`, instead of sleeping out the timeout. That turns five
   minutes of nothing into a one-line, correctly-attributed failure — and it helps every other
   cause of a non-ready runtime too, including the Hugging Face one.
3. **Scope `pkill` to this job's process** once ports are per-job, so cleanup cannot reach across
   to a sibling.

### What this does not weaken

No timeout is raised, no assertion relaxed, no test retried or skipped, no required check removed.
Item 1 removes a shared-resource collision, item 2 only adds information to a failure that already
exists, item 3 narrows a `pkill` that is currently host-wide.

## Note on filing

Filed rather than PR'd: the account running this sweep (`grokspice`) has no push access to
`spiceai/spiceai`. It also cannot apply labels (`HTTP 403 Must have admin rights to Repository`)
or set an assignee (not in the repo's assignable list). Suggested labels: `area/ci`, `kind/bug`.

<!-- harden-ci-workflows: e2e-fixed-port-collision-shared-macos-pool -->

--- a/.github/actions/allocate-spice-ports/action.yml
+++ b/.github/actions/allocate-spice-ports/action.yml
@@ -1,0 +1,12 @@
+name: 'Allocate Spice runtime ports'
+description: |
+  Derives an HTTP and a Flight port for this job's Spice runtime and exports
+  them as SPICE_HTTP_PORT / SPICE_FLIGHT_PORT / SPICE_HTTP_ENDPOINT /
+  SPICE_FLIGHT_ENDPOINT, so jobs sharing a self-hosted host cannot collide on
+  spiced's defaults 8090 and 50051 (#12419).
+runs:
+  using: 'composite'
+  steps:
+    - name: Allocate Spice runtime ports
+      shell: bash
+      run: bash "${GITHUB_ACTION_PATH}/allocate_ports.sh"

--- a/.github/actions/allocate-spice-ports/allocate_ports.sh
+++ b/.github/actions/allocate-spice-ports/allocate_ports.sh
@@ -80,25 +80,28 @@ if [ -z "${hash_value}" ]; then
 fi
 
 offset=$((hash_value % SPAN))
-http_port=$((HTTP_BASE + offset))
 
 # Walk both ports together so the pair keeps a constant distance; that way one
 # number in a log identifies the other, and neither walks into the other's
-# window.
+# window. The step wraps within SPAN rather than running past it, so a seed that
+# lands near the top of the window cannot walk out of the range this script
+# promises — and into the ephemeral range it was chosen to avoid.
 probe=0
-while [ "${probe}" -lt "${MAX_PROBE}" ]; do
+while :; do
+  http_port=$((HTTP_BASE + (offset + probe) % SPAN))
   flight_port=$((http_port + FLIGHT_OFFSET))
 
   if ! port_in_use "${http_port}" && ! port_in_use "${flight_port}"; then
     break
   fi
 
-  echo "Port ${http_port}/${flight_port} is already held on this host; trying the next pair."
-  http_port=$((http_port + 1))
   probe=$((probe + 1))
-done
+  if [ "${probe}" -ge "${MAX_PROBE}" ]; then
+    break
+  fi
 
-flight_port=$((http_port + FLIGHT_OFFSET))
+  echo "Port ${http_port}/${flight_port} is already held on this host; trying the next pair."
+done
 
 if [ "${probe}" -ge "${MAX_PROBE}" ]; then
   # Report rather than fail: spiced will refuse to bind and say so, and the

--- a/.github/actions/allocate-spice-ports/allocate_ports.sh
+++ b/.github/actions/allocate-spice-ports/allocate_ports.sh
@@ -1,0 +1,129 @@
+#!/usr/bin/env bash
+#
+# Pick the HTTP and Flight ports this job's Spice runtime should bind, and
+# export them for every later step.
+#
+# Why a job needs its own ports at all: the self-hosted pools run several runner
+# instances per physical host (`/opt/github-runner-01`, `-02`, `-03`, … on one
+# machine), and those instances share a network namespace. Every `spice run`
+# binds spiced's defaults, 127.0.0.1:8090 and 127.0.0.1:50051, so two jobs that
+# overlap on one host cannot both bind. The second exits, and the readiness poll
+# then waits out its whole timeout with nothing to show for it. Worse, if the
+# process already holding 8090 is a *ready* runtime from the other job, the poll
+# succeeds and the tests run against a spicepod they did not load. See #12419.
+#
+# The port is derived from the runner instance rather than from the run or the
+# job, because the runner instance is the unit that shares the host: one
+# instance runs one job at a time, so distinct instances holding distinct ports
+# is exactly the property that makes the collision impossible. Run and job are
+# folded in as well so that a port leaked by an earlier job on this same
+# instance is unlikely to be the one we ask for next.
+#
+# Outputs, written to $GITHUB_ENV:
+#   SPICE_HTTP_PORT       e.g. 21734
+#   SPICE_FLIGHT_PORT     e.g. 26734
+#   SPICE_HTTP_ENDPOINT   e.g. http://127.0.0.1:21734
+#   SPICE_FLIGHT_ENDPOINT e.g. 127.0.0.1:26734   (bare host:port — `spiced --flight` takes no scheme)
+
+set -uo pipefail
+
+# Both windows sit below the ephemeral range on every runner OS — Linux
+# allocates from 32768 and macOS from 49152 — so a port picked here cannot be
+# taken from under us by an unrelated outbound connection. They are also far
+# from spiced's own defaults, which keeps a stray 8090 in a log unambiguous:
+# that is a step that has not been given the job's port.
+HTTP_BASE="${SPICE_PORT_HTTP_BASE:-20000}"
+FLIGHT_OFFSET="${SPICE_PORT_FLIGHT_OFFSET:-5000}"
+SPAN="${SPICE_PORT_SPAN:-4000}"
+# How far to walk forward when the derived port is already held. Small on
+# purpose: a busy derived port means a leak or a hash collision, and walking a
+# long way would only hide it.
+MAX_PROBE="${SPICE_PORT_MAX_PROBE:-25}"
+
+# Seed. RUNNER_NAME is the one that matters — it identifies the runner instance,
+# and therefore the collision domain. The others only spread repeat jobs on one
+# instance apart. All are optional so the script is runnable outside Actions.
+seed_source="${RUNNER_NAME:-}|${GITHUB_RUN_ID:-}|${GITHUB_JOB:-}|${GITHUB_RUN_ATTEMPT:-}"
+
+# `cksum` rather than a shell hash or `md5`: it is POSIX, present on both runner
+# families, and its first field is already a decimal integer.
+seed_hash() {
+  printf '%s' "$1" | cksum | awk '{print $1}'
+}
+
+# True when something is already listening on the port. Tries each tool the
+# runners might have and, if none is present, answers "not in use" — a probe we
+# cannot run must not block the job, and binding is still checked for real by
+# spiced a moment later.
+port_in_use() {
+  local port="$1"
+
+  if command -v lsof >/dev/null 2>&1; then
+    lsof -nP -iTCP:"${port}" -sTCP:LISTEN >/dev/null 2>&1
+    return $?
+  fi
+
+  if command -v netstat >/dev/null 2>&1; then
+    # `.port` is the macOS/BSD separator, `:port` the Linux one.
+    netstat -an 2>/dev/null | grep -E "[.:]${port}[[:space:]]+.*LISTEN" >/dev/null 2>&1
+    return $?
+  fi
+
+  return 1
+}
+
+hash_value="$(seed_hash "${seed_source}")"
+if [ -z "${hash_value}" ]; then
+  # cksum is missing or produced nothing. $$ is unique among live processes on
+  # the host, which is the same property the seed was after.
+  hash_value=$$
+fi
+
+offset=$((hash_value % SPAN))
+http_port=$((HTTP_BASE + offset))
+
+# Walk both ports together so the pair keeps a constant distance; that way one
+# number in a log identifies the other, and neither walks into the other's
+# window.
+probe=0
+while [ "${probe}" -lt "${MAX_PROBE}" ]; do
+  flight_port=$((http_port + FLIGHT_OFFSET))
+
+  if ! port_in_use "${http_port}" && ! port_in_use "${flight_port}"; then
+    break
+  fi
+
+  echo "Port ${http_port}/${flight_port} is already held on this host; trying the next pair."
+  http_port=$((http_port + 1))
+  probe=$((probe + 1))
+done
+
+flight_port=$((http_port + FLIGHT_OFFSET))
+
+if [ "${probe}" -ge "${MAX_PROBE}" ]; then
+  # Report rather than fail: spiced will refuse to bind and say so, and the
+  # readiness wait turns that into a named failure. Failing here instead would
+  # replace a specific diagnosis with a vaguer one.
+  echo "::warning::Could not find a free port pair within ${MAX_PROBE} tries of ${HTTP_BASE}+${offset}; using ${http_port}/${flight_port} anyway."
+fi
+
+http_endpoint="http://127.0.0.1:${http_port}"
+# No scheme: `spice run --flight-endpoint` hands this to `spiced --flight`
+# verbatim, and that flag takes a bind address.
+flight_endpoint="127.0.0.1:${flight_port}"
+
+echo "Spice runtime ports for this job: HTTP ${http_port}, Flight ${flight_port} (runner '${RUNNER_NAME:-unknown}')."
+
+assignments="SPICE_HTTP_PORT=${http_port}
+SPICE_FLIGHT_PORT=${flight_port}
+SPICE_HTTP_ENDPOINT=${http_endpoint}
+SPICE_FLIGHT_ENDPOINT=${flight_endpoint}"
+
+# Always logged: whoever reads a failed job needs the port to make sense of the
+# readiness poll and of stop-spice's report. Outside Actions this is also the
+# only output, so the script stays runnable by hand.
+printf '%s\n' "${assignments}"
+
+if [ -n "${GITHUB_ENV:-}" ]; then
+  printf '%s\n' "${assignments}" >>"${GITHUB_ENV}"
+fi

--- a/.github/actions/stop-spice/action.yml
+++ b/.github/actions/stop-spice/action.yml
@@ -10,9 +10,12 @@ inputs:
     required: false
     default: '.'
   ports:
-    description: 'Space-separated TCP ports to wait for release'
+    description: |
+      Space-separated TCP ports to wait for release. Empty lets the script use
+      the ports this job allocated (SPICE_HTTP_PORT / SPICE_FLIGHT_PORT), and
+      spiced's defaults when it allocated none.
     required: false
-    default: '8090 50051'
+    default: ''
   print-log:
     description: 'Print spice.log after stopping'
     required: false

--- a/.github/actions/stop-spice/stop_spice.sh
+++ b/.github/actions/stop-spice/stop_spice.sh
@@ -35,7 +35,13 @@ set -uo pipefail
 
 # Ports `spiced` binds (HTTP, Flight). Left unquoted where used so the
 # space-separated value splits into words.
-SPICE_STOP_PORTS="${SPICE_STOP_PORTS:-8090 50051}"
+#
+# A job that allocated its own ports exports them (see
+# .github/actions/allocate-spice-ports, #12419); check those rather than the
+# defaults, or this reports on ports belonging to a *different* runner
+# instance's job — noise that reads as a leak — while never confirming ours
+# came free. Jobs that did not allocate keep spiced's defaults.
+SPICE_STOP_PORTS="${SPICE_STOP_PORTS:-${SPICE_HTTP_PORT:-8090} ${SPICE_FLIGHT_PORT:-50051}}"
 
 # Seconds to wait for a graceful exit before escalating to SIGKILL, and then for
 # the ports to clear. Kept short: this runs in cleanup on every job.

--- a/.github/workflows/e2e_lifecycle_scripts_test.yml
+++ b/.github/workflows/e2e_lifecycle_scripts_test.yml
@@ -15,17 +15,23 @@ on:
       - release/*
       - feature-*
     paths:
+      - '.github/actions/allocate-spice-ports/**'
       - '.github/actions/stop-spice/**'
       - '.github/actions/wait-for-spice-ready/**'
       - 'scripts/test_e2e_spice_lifecycle.sh'
+      - 'scripts/test_e2e_spice_ports.sh'
+      - 'test/models/**'
       - '.github/workflows/e2e_lifecycle_scripts_test.yml'
   push:
     branches:
       - trunk
     paths:
+      - '.github/actions/allocate-spice-ports/**'
       - '.github/actions/stop-spice/**'
       - '.github/actions/wait-for-spice-ready/**'
       - 'scripts/test_e2e_spice_lifecycle.sh'
+      - 'scripts/test_e2e_spice_ports.sh'
+      - 'test/models/**'
       - '.github/workflows/e2e_lifecycle_scripts_test.yml'
   workflow_dispatch:
 
@@ -62,3 +68,28 @@ jobs:
         # Invoked through `bash` so the job does not depend on the script's
         # executable bit surviving a checkout on every platform.
         run: bash scripts/test_e2e_spice_lifecycle.sh
+
+      - name: Run port allocation tests
+        run: bash scripts/test_e2e_spice_ports.sh
+
+      # The REPL scripts and their tests only ran inside the model E2E jobs,
+      # which are skipped on pull requests — so a change to them reached the
+      # merge queue unexercised. They need nothing but `expect`, so they can run
+      # here on every PR that touches them instead.
+      - name: Install expect (linux)
+        if: matrix.os == 'linux-x64'
+        run: |
+          sudo apt-get update
+          sudo apt-get install -y expect
+
+      - name: Install expect (macOS)
+        if: matrix.os == 'macos-arm64'
+        run: |
+          if command -v expect >/dev/null 2>&1; then
+            expect -v
+          else
+            HOMEBREW_NO_AUTO_UPDATE=1 HOMEBREW_NO_ENV_HINTS=1 HOMEBREW_NO_INSTALL_CLEANUP=1 brew install expect
+          fi
+
+      - name: Run REPL script tests
+        run: bash test/models/expect_test.sh

--- a/.github/workflows/e2e_test_ci.yml
+++ b/.github/workflows/e2e_test_ci.yml
@@ -267,16 +267,23 @@ jobs:
           cp ./test/models/spicepod_openai.yml ./spicepod.yaml
           cat ./spicepod.yaml
 
+      # This job runs on the self-hosted pool, where several runner instances
+      # share one host and therefore one network namespace, so spiced's default
+      # 8090/50051 cannot be assumed free (#12419).
+      - name: Allocate Spice runtime ports
+        uses: ./.github/actions/allocate-spice-ports
+
       - name: Start spice runtime
         env:
           GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
           SPICE_OPENAI_API_KEY: ${{ secrets.SPICE_SECRET_OPENAI_API_KEY }}
         run: |
-          spice run &> spice.log &
+          spice run --http-endpoint "${SPICE_HTTP_ENDPOINT}" --flight-endpoint "${SPICE_FLIGHT_ENDPOINT}" &> spice.log &
 
       - name: Wait for Spice runtime is ready
         uses: ./.github/actions/wait-for-spice-ready
         with:
+          url: ${{ env.SPICE_HTTP_ENDPOINT }}/v1/ready
           timeout: '60'
 
       - name: Install expect (linux)
@@ -463,16 +470,21 @@ jobs:
 
           exit "${rc}"
 
+      # Same shared-host constraint as the OpenAI model job above (#12419).
+      - name: Allocate Spice runtime ports
+        uses: ./.github/actions/allocate-spice-ports
+
       - name: Start spice runtime
         env:
           GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
           SPICE_OPENAI_API_KEY: ${{ secrets.SPICE_SECRET_OPENAI_API_KEY }}
         run: |
-          spice run &> spice.log &
+          spice run --http-endpoint "${SPICE_HTTP_ENDPOINT}" --flight-endpoint "${SPICE_FLIGHT_ENDPOINT}" &> spice.log &
 
       - name: Wait for Spice runtime is ready
         uses: ./.github/actions/wait-for-spice-ready
         with:
+          url: ${{ env.SPICE_HTTP_ENDPOINT }}/v1/ready
           timeout: '300'
 
       - name: Install expect (linux)

--- a/scripts/test_e2e_spice_ports.sh
+++ b/scripts/test_e2e_spice_ports.sh
@@ -1,0 +1,250 @@
+#!/usr/bin/env bash
+#
+# Tests for .github/actions/allocate-spice-ports/allocate_ports.sh — the script
+# that gives each E2E job its own spiced ports so jobs sharing a self-hosted
+# host cannot collide on 8090/50051 (#12419).
+#
+# What matters here is not the specific number it picks but the properties the
+# collision fix rests on: two runner instances must not be handed the same pair,
+# the same instance must be handed the same pair every time, the pair must stay
+# clear of the OS ephemeral range, and a port that is already held must be
+# stepped over rather than handed out.
+#
+#   bash scripts/test_e2e_spice_ports.sh
+
+set -uo pipefail
+
+repo_root=$(cd -- "$(dirname -- "$0")/.." && pwd)
+script="$repo_root/.github/actions/allocate-spice-ports/allocate_ports.sh"
+
+work_dir=$(mktemp -d)
+trap 'rm -rf "$work_dir"' EXIT
+
+failures=0
+
+pass() {
+  printf '  ok    %s\n' "$1"
+}
+
+fail() {
+  printf '  FAIL  %s\n' "$1"
+  failures=$((failures + 1))
+}
+
+assert_eq() {
+  # assert_eq <what> <expected> <actual>
+  if [ "$2" = "$3" ]; then
+    pass "$1"
+  else
+    fail "$1: expected '$2', got '$3'"
+  fi
+}
+
+assert_ne() {
+  # assert_ne <what> <a> <b>
+  if [ "$2" != "$3" ]; then
+    pass "$1"
+  else
+    fail "$1: both were '$2'"
+  fi
+}
+
+assert_between() {
+  # assert_between <what> <low> <high> <value>
+  if [ "$4" -ge "$2" ] && [ "$4" -le "$3" ]; then
+    pass "$1"
+  else
+    fail "$1: $4 is outside $2..$3"
+  fi
+}
+
+# run_alloc <runner-name> [extra env assignments...] — runs the script with a
+# fresh GITHUB_ENV and leaves the resulting assignments in $alloc_env, the
+# stdout in $alloc_output.
+alloc_env=""
+alloc_output=""
+run_alloc() {
+  local runner="$1"
+  shift
+
+  local env_file="$work_dir/github_env"
+  : >"$env_file"
+
+  alloc_output=$(env RUNNER_NAME="$runner" GITHUB_ENV="$env_file" "$@" bash "$script" 2>&1)
+  alloc_env=$(cat "$env_file")
+}
+
+# value_of <key> — reads a KEY=VALUE out of the last run's GITHUB_ENV.
+value_of() {
+  printf '%s\n' "$alloc_env" | sed -n "s/^$1=//p"
+}
+
+printf 'case: exports the four variables a job needs\n'
+run_alloc 'github-runner-01'
+for key in SPICE_HTTP_PORT SPICE_FLIGHT_PORT SPICE_HTTP_ENDPOINT SPICE_FLIGHT_ENDPOINT; do
+  if [ -n "$(value_of "$key")" ]; then
+    pass "exports $key"
+  else
+    fail "did not export $key"
+  fi
+done
+
+printf 'case: the endpoints agree with the ports\n'
+http_port=$(value_of SPICE_HTTP_PORT)
+flight_port=$(value_of SPICE_FLIGHT_PORT)
+assert_eq 'HTTP endpoint carries the http:// scheme and the HTTP port' \
+  "http://127.0.0.1:${http_port}" "$(value_of SPICE_HTTP_ENDPOINT)"
+# `spice run --flight-endpoint` is handed to `spiced --flight` verbatim, and
+# that flag takes a bind address. A scheme here would reach spiced as part of
+# the address and fail the bind.
+assert_eq 'Flight endpoint is a bare host:port, with no scheme' \
+  "127.0.0.1:${flight_port}" "$(value_of SPICE_FLIGHT_ENDPOINT)"
+
+printf 'case: ports avoid spiced defaults and the OS ephemeral range\n'
+# Linux allocates ephemeral ports from 32768 and macOS from 49152, so staying
+# under 32768 keeps an unrelated outbound connection from taking the port first.
+assert_between 'HTTP port sits in the reserved window' 20000 23999 "$http_port"
+assert_between 'Flight port sits in the reserved window' 25000 28999 "$flight_port"
+assert_ne 'HTTP port is not spiced default 8090' '8090' "$http_port"
+assert_ne 'Flight port is not spiced default 50051' '50051' "$flight_port"
+
+printf 'case: the same runner instance is handed the same pair every time\n'
+run_alloc 'github-runner-01' GITHUB_RUN_ID=100 GITHUB_JOB=test_openai_model
+first_http=$(value_of SPICE_HTTP_PORT)
+run_alloc 'github-runner-01' GITHUB_RUN_ID=100 GITHUB_JOB=test_openai_model
+assert_eq 'a repeated allocation is stable' "$first_http" "$(value_of SPICE_HTTP_PORT)"
+
+printf 'case: runner instances sharing a host get different pairs\n'
+# This is the whole point: -01 and -02 live on one machine and share its network
+# namespace, so identical ports here would reproduce the bug being fixed.
+declare -a seen=()
+collisions=0
+for instance in 01 02 03 04 05 06 07 08; do
+  run_alloc "github-runner-${instance}" GITHUB_RUN_ID=100 GITHUB_JOB=test_openai_model
+  port=$(value_of SPICE_HTTP_PORT)
+  for previous in ${seen[@]+"${seen[@]}"}; do
+    if [ "$previous" = "$port" ]; then
+      collisions=$((collisions + 1))
+    fi
+  done
+  seen+=("$port")
+done
+assert_eq 'eight runner instances on one host get eight distinct ports' '0' "$collisions"
+
+printf 'case: a held port is stepped over\n'
+# Hold the port this seed would otherwise pick, then ask again: the script must
+# hand out a different one rather than a port spiced cannot bind.
+run_alloc 'github-runner-holder' GITHUB_RUN_ID=200 GITHUB_JOB=test_hf_model
+wanted_http=$(value_of SPICE_HTTP_PORT)
+
+# A listener that stays up for the duration of the probe. `nc -l` differs
+# between the BSD and GNU builds, so bind with whatever the runner has.
+holder_pid=''
+if command -v python3 >/dev/null 2>&1; then
+  python3 - "$wanted_http" <<'PY' &
+import socket, sys, time
+s = socket.socket()
+s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+s.bind(("127.0.0.1", int(sys.argv[1])))
+s.listen(1)
+time.sleep(30)
+PY
+  holder_pid=$!
+fi
+
+if [ -n "$holder_pid" ]; then
+  # Give the listener a moment to be visible to lsof/netstat.
+  sleep 1
+
+  if command -v lsof >/dev/null 2>&1 || command -v netstat >/dev/null 2>&1; then
+    run_alloc 'github-runner-holder' GITHUB_RUN_ID=200 GITHUB_JOB=test_hf_model
+    assert_ne 'a held port is not handed out' "$wanted_http" "$(value_of SPICE_HTTP_PORT)"
+    case $alloc_output in
+    *'already held'*) pass 'says why it moved on' ;;
+    *) fail "output does not mention the held port: $alloc_output" ;;
+    esac
+  else
+    printf '  skip  no lsof or netstat on this host, cannot test the probe\n'
+  fi
+
+  kill "$holder_pid" 2>/dev/null
+  wait "$holder_pid" 2>/dev/null
+else
+  printf '  skip  no python3 to hold a port with\n'
+fi
+
+# ---------------------------------------------------------------------------
+# stop-spice checks the ports this job actually used
+# ---------------------------------------------------------------------------
+#
+# Cleanup waits for "the ports" to come free. Once a job binds its own pair,
+# waiting on 8090/50051 instead would watch another runner instance's job — and
+# never confirm ours was released. The full stop-spice behaviour is covered by
+# scripts/test_e2e_spice_lifecycle.sh; this covers only which ports it picks.
+
+stop_script="$repo_root/.github/actions/stop-spice/stop_spice.sh"
+
+# Run in an empty directory with no spice processes to find, so the script falls
+# straight through to its port report. `RUNNER_WORKSPACE` scopes it to a path
+# that owns nothing, which keeps it from touching anything on this machine.
+run_stop() {
+  local empty="$work_dir/empty"
+  mkdir -p "$empty"
+  (cd "$empty" && env RUNNER_WORKSPACE="$empty" SPICE_STOP_PORT_WAIT=0 SPICE_STOP_GRACE=0 \
+    "$@" bash "$stop_script" 2>&1)
+}
+
+printf 'case: stop-spice checks the ports the job allocated\n'
+stop_output=$(run_stop SPICE_HTTP_PORT=21734 SPICE_FLIGHT_PORT=26734)
+case $stop_output in
+*':21734'*) pass 'checks the allocated HTTP port' ;;
+*) fail "does not mention :21734: $stop_output" ;;
+esac
+case $stop_output in
+*':26734'*) pass 'checks the allocated Flight port' ;;
+*) fail "does not mention :26734: $stop_output" ;;
+esac
+case $stop_output in
+*':8090'*) fail "still checks the default :8090: $stop_output" ;;
+*) pass 'does not check the default :8090' ;;
+esac
+
+printf 'case: stop-spice keeps the defaults when no ports were allocated\n'
+stop_output=$(run_stop)
+case $stop_output in
+*':8090'*) pass 'falls back to the default HTTP port' ;;
+*) fail "does not mention :8090: $stop_output" ;;
+esac
+case $stop_output in
+*':50051'*) pass 'falls back to the default Flight port' ;;
+*) fail "does not mention :50051: $stop_output" ;;
+esac
+
+printf 'case: an explicit port list still wins\n'
+stop_output=$(run_stop SPICE_HTTP_PORT=21734 SPICE_FLIGHT_PORT=26734 SPICE_STOP_PORTS=9999)
+case $stop_output in
+*':9999'*) pass 'honours SPICE_STOP_PORTS' ;;
+*) fail "does not mention :9999: $stop_output" ;;
+esac
+case $stop_output in
+*':21734'*) fail "explicit list did not override the allocated ports: $stop_output" ;;
+*) pass 'ignores the allocated ports when told which to check' ;;
+esac
+
+printf 'case: works with no Actions environment at all\n'
+# Runnable by hand, and by anyone reproducing a CI failure locally.
+bare_output=$(env -u RUNNER_NAME -u GITHUB_ENV -u GITHUB_RUN_ID -u GITHUB_JOB \
+  -u GITHUB_RUN_ATTEMPT bash "$script" 2>&1)
+bare_status=$?
+assert_eq 'exits 0 with no Actions variables set' '0' "$bare_status"
+case $bare_output in
+*SPICE_HTTP_PORT=*) pass 'prints the assignments when GITHUB_ENV is unset' ;;
+*) fail "no assignments in output: $bare_output" ;;
+esac
+
+if [ "$failures" -ne 0 ]; then
+  printf '\n%s check(s) failed\n' "$failures"
+  exit 1
+fi
+
+printf '\nAll checks passed\n'

--- a/scripts/test_e2e_spice_ports.sh
+++ b/scripts/test_e2e_spice_ports.sh
@@ -194,6 +194,21 @@ run_stop() {
     "$@" bash "$stop_script" 2>&1)
 }
 
+printf 'case: probing from the top of the window wraps instead of walking out of it\n'
+# A seed landing on the last port in the window, with that port held, must wrap
+# to the bottom rather than step to 24000 — which would leave the reserved
+# window and head toward the ephemeral range the window exists to avoid.
+run_alloc 'github-runner-01' SPICE_PORT_SPAN=1 SPICE_PORT_MAX_PROBE=3
+assert_eq 'a span of one always yields the base port' \
+  '20000' "$(value_of SPICE_HTTP_PORT)"
+
+# Ten probes from the top of a small window stay inside it.
+for probe_seed in 1 2 3 4 5 6 7 8; do
+  run_alloc "github-runner-${probe_seed}" SPICE_PORT_SPAN=4 SPICE_PORT_MAX_PROBE=10
+  assert_between "a span of four keeps seed ${probe_seed} in the window" \
+    20000 20003 "$(value_of SPICE_HTTP_PORT)"
+done
+
 printf 'case: stop-spice checks the ports the job allocated\n'
 stop_output=$(run_stop SPICE_HTTP_PORT=21734 SPICE_FLIGHT_PORT=26734)
 case $stop_output in

--- a/test/models/chat_01.exp
+++ b/test/models/chat_01.exp
@@ -5,7 +5,7 @@ source [file join [file dirname [info script]] repl_helpers.exp]
 # Runtime response timeout
 set timeout 30
 
-spawn spice chat
+spawn spice chat {*}[repl_endpoint_args]
 
 proc verify_datasets_access {} {
     global timeout

--- a/test/models/chat_01_simple.exp
+++ b/test/models/chat_01_simple.exp
@@ -5,7 +5,7 @@ source [file join [file dirname [info script]] repl_helpers.exp]
 # Runtime response timeout
 set timeout 30
 
-spawn spice chat
+spawn spice chat {*}[repl_endpoint_args]
 
 proc chat {message}  {
     global timeout

--- a/test/models/expect_test.sh
+++ b/test/models/expect_test.sh
@@ -220,6 +220,12 @@ mode=$1
 exit_before=${SPICE_FAKE_EXIT_BEFORE_TURN:-0}
 exit_after=${SPICE_FAKE_EXIT_AFTER_TURN:-0}
 
+# Records how the script invoked us, so a test can check that the runtime
+# endpoint was passed through rather than left at the CLI default.
+if [ -n "${SPICE_FAKE_ARGV_FILE:-}" ]; then
+  printf '%s\n' "$*" >"$SPICE_FAKE_ARGV_FILE"
+fi
+
 prompt='chat> '
 if [ "$mode" = 'search' ]; then
   prompt='search> '
@@ -303,6 +309,48 @@ script_case 'chat_01.exp when the REPL exits while idle' chat_01.exp \
 assert_status 1
 assert_reports 'Checking the chat REPL is still running'
 assert_reports 'exited with status 44'
+
+# ---------------------------------------------------------------------------
+# The runtime endpoint the REPL is pointed at
+# ---------------------------------------------------------------------------
+#
+# In CI each job binds its runtime to its own port (#12419), so a REPL left on
+# the CLI default would talk to whatever else holds 8090 on a shared host — or
+# to nothing. These check the scripts forward the endpoint when it is set, and
+# leave the default alone when it is not.
+
+argv_file="$work_dir/spice_argv"
+
+assert_argv() {
+  local expected=$1 actual
+  actual=$(cat "$argv_file" 2>/dev/null)
+  if [ "$actual" = "$expected" ]; then
+    pass "invokes spice as \"$expected\""
+  else
+    fail "expected spice args \"$expected\", got \"$actual\""
+  fi
+}
+
+for script in chat_01.exp chat_01_simple.exp search_01.exp; do
+  subcommand=chat
+  case $script in
+  search_*) subcommand=search ;;
+  esac
+
+  : >"$argv_file"
+  script_case "$script forwards SPICE_HTTP_ENDPOINT" "$script" \
+    SPICE_FAKE_ARGV_FILE="$argv_file" \
+    SPICE_HTTP_ENDPOINT='http://127.0.0.1:21734'
+  assert_status 0
+  assert_argv "$subcommand --endpoint http://127.0.0.1:21734"
+
+  : >"$argv_file"
+  script_case "$script keeps the CLI default when SPICE_HTTP_ENDPOINT is unset" "$script" \
+    SPICE_FAKE_ARGV_FILE="$argv_file" \
+    SPICE_HTTP_ENDPOINT=''
+  assert_status 0
+  assert_argv "$subcommand"
+done
 
 if [ "$failures" -ne 0 ]; then
   printf '\n%s check(s) failed\n' "$failures"

--- a/test/models/repl_helpers.exp
+++ b/test/models/repl_helpers.exp
@@ -70,6 +70,21 @@ proc repl_send {text} {
     }
 }
 
+# The arguments that point a REPL at this job's runtime, or none at all.
+#
+# In CI each job binds its runtime to its own port, because the self-hosted
+# pools run several runner instances on one host and spiced's default 8090
+# cannot be shared (#12419). SPICE_HTTP_ENDPOINT carries that port. When it is
+# unset — running one of these scripts by hand against a local runtime — the
+# REPL keeps its own default, so nothing has to be set to reproduce a CI run.
+proc repl_endpoint_args {} {
+    if {[info exists ::env(SPICE_HTTP_ENDPOINT)] && $::env(SPICE_HTTP_ENDPOINT) ne ""} {
+        return [list --endpoint $::env(SPICE_HTTP_ENDPOINT)]
+    }
+
+    return [list]
+}
+
 # Fails the script if the spawned process has exited. Writing to the pty of an
 # exited process succeeds silently, so a REPL that crashes while sitting idle at
 # its prompt is invisible to `repl_send` alone; reading finds the eof. Output

--- a/test/models/search_01.exp
+++ b/test/models/search_01.exp
@@ -5,7 +5,7 @@ source [file join [file dirname [info script]] repl_helpers.exp]
 # spiced runtime response timeout
 set timeout 5
 
-spawn spice search
+spawn spice search {*}[repl_endpoint_args]
 
 proc perform_search {search_term expected_term} {
     global timeout

